src/Raven.Server/Documents/Replication/DocumentReplicationExecuter.cs

```diff
@@ -1,51 +1,92 @@
+using System.IO;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using Raven.Abstractions.Data;
+using Raven.Abstractions.Replication;
+using Raven.Client.Linq;
+using Raven.Imports.Newtonsoft.Json.Utilities;
 using Raven.Server.ReplicationUtil;
 using Raven.Server.ServerWide.Context;
 using Sparrow.Json;
 
 namespace Raven.Server.Documents.Replication
 {
     //TODO : add DocumentReplicationStatistics that will track operational data
+    //TODO : do not forget to handle authentication stuff
     public class DocumentReplicationExecuter: BaseReplicationExecuter
     {
         private readonly ReplicationWebSocket _socket;
-        private readonly DocumentReplicationDestination _config;
+        private readonly Guid _srcDbId;
+        private readonly ReplicationDestination _destination;
+        private long _lastSentEtag;
 
-        public DocumentReplicationExecuter(DocumentDatabase database, DocumentReplicationDestination config) : base(database)
+        public DocumentReplicationExecuter(DocumentDatabase database, 
+            Guid srcDbId, 
+            ReplicationDestination destination, 
+            long lastSentEtag) : base(database)
         {
-            _config = config;
-            _socket = new ReplicationWebSocket(config.Url);
+            _srcDbId = srcDbId;
+            _destination = destination;
+            _lastSentEtag = lastSentEtag;
+            _socket = new ReplicationWebSocket(destination.Url);		
         }
 
-        public override string Name => _config.Url;
+        //if _destination == null --> the replication is only incoming, otherwise it is a two-way one
+        public override string Name => _srcDbId + (_destination != null ? $"/{_destination.Url}" : string.Empty);
 
         //by design this method won't handle opening and commit of the transaction
+        //(that should happen at the calling code)
         public void ReceiveReplicatedDocuments(DocumentsOperationContext context,
-            IEnumerable<BlittableJsonReaderObject> docs)
+            List<BlittableJsonReaderObject> docs)
         {
-            //TODO: finish this
             var dbChangeVector = _database.DocumentsStorage.GetChangeVector(context);
-            foreach (var doc in docs)
+            var changeVectorUpdated = false;
+            for(int i = 0; i < docs.Count; i++)
             {
-                UpdateDbChangeVectorIfNeededFrom(doc);
+                var doc = docs[i];
+                changeVectorUpdated = UpdateDbChangeVectorIfNeededFrom(doc,dbChangeVector);
                 ReceiveReplicated(context, doc);
             }
+
+            if(changeVectorUpdated)
+                _database.DocumentsStorage.SetChangeVector(context,dbChangeVector);
         }
 
-        private void UpdateDbChangeVectorIfNeededFrom(BlittableJsonReaderObject doc)
+        private bool UpdateDbChangeVectorIfNeededFrom(BlittableJsonReaderObject doc,ChangeVectorEntry[] dbChangeVector)
```

Comment:
Don't do it like this, you are forcing us to do a LOT of checks on every round.
Instead, have a single dictionary per batch that will get the max of all the items, then update the db change vector once.